pydrofoil/real.py

```diff
@@ -0,0 +1,264 @@
+    #     # dif = len(s)-1-i
+    #     # if dif < 19:
+    #     #     den = rbigint.fromint(10**dif)
+    #     # else:
+    #     #     den = den_of_fromstr(dif)
+    #     # # return Real(num, den)
+    #     # return lead_zero_count, num, den.str()
+
+
+
+    def add(self, other):
+        num_new_1 = self.num.mul(other.den)
+        num_new_2 = other.num.mul(self.den)
+        den_new = self.den.mul(other.den)
+        num_new = num_new_1.add(num_new_2)
+        return Real(num_new, den_new)
+    
+    def sub(self, other):
+        num_new_1 = self.num.mul(other.den)
+        num_new_2 = other.num.mul(self.den)
+        den_new = self.den.mul(other.den)
+        num_new = num_new_1.sub(num_new_2)
+        return Real(num_new, den_new)
+    
+    def mul(self, other):
+        num_new = self.num.mul(other.num)
+        den_new = self.den.mul(other.den)
+        return Real(num_new, den_new)
+    
+    def div(self, other):
+        if self.num.int_eq(0) and not other.num.int_eq(0):
+            num_new = rbigint.fromint(0)
+            den_new = rbigint.fromint(1)
+        elif other.num.int_eq(0):
+            assert False, "ZerodivideError: denominator cannot be 0"
+        else:
+            num_new = self.num.mul(other.den)
+            den_new = self.den.mul(other.num)
+        if den_new.int_lt(0):
+            num_new = num_new.neg()
+            den_new = den_new.neg()
+        return Real(num_new, den_new)
+
+    def pow(self, n):
+        if isinstance(n, int):
+            if n == 0:
+                return Real(rbigint.fromint(1), rbigint.fromint(1))
+            elif n < MININT or n > MAXINT:
+                assert False, "exponent is out of range of INT"
+            else:
+                n = rbigint.fromint(n)
+                num_new = self.num.pow(n)
+                den_new = self.den.pow(n)
+                return Real(num_new, den_new)
+        elif isinstance(n, rbigint):
+            if n.int_eq(0):
+                return Real(rbigint.fromint(1), rbigint.fromint(1))
+            else:
+                n = rbigint.fromint(n)
+                num_new = self.num.pow(n)
+                den_new = self.den.pow(n)
+                return Real(num_new, den_new)
+        elif isinstance(n, Real):
+            if n.num.int_eq(0):
+                return Real(rbigint.fromint(1), rbigint.fromint(1))
+            else:
+                if n.den.int_eq(1):
+                    num_new = self.num.pow(n.num)
+                    den_new = self.den.pow(n.num)
+                    return Real(num_new, den_new)
+                else:
+                    assert False, "exponent cannot be fraction"
+        else:
+            raise TypeError("exponent doesn't support this type")
+        
+        
+            
+    
+    def neg(self):
+        return Real(self.num.neg(), self.den)
+    
+    def abs(self):
+        return Real(self.num.abs(), self.den)
+
+    def ceil(self):
```

Comment:
`ceil` and `floor` have the following signatures:

```
val floor = pure "round_down" : real -> int
val ceil = pure "round_up" : real -> int
```

that means they should return an `rbigint`.

for `floor` you should be able to use `rbigint.floordiv(self.num, self.den)`. for `ceil` you could do `self.num.neg.floordiv(self.den).neg()` I think?

Reply:
ok, I'm going to check it